app.py

```diff
@@ -113,109 +16,109 @@ def llm_prompt(style, n_words):
+            return ui.card(
+                ui.card_header(
+                    {"class": "bg-dark fw-bold fs-3"},
+                    ui.output_markdown_stream("card_title")
+                ),
+                ui.output_markdown_stream("card_body", auto_scroll=False),
+            )
+
+    # As .append_message_stream() is iterating through the generator, accumulate contents
+    # so we can parse it, extract the image details, and update the card title and body.
+    async def stream_wrapper(stream):
+        all_content: str = ""
+        async with card_body.stream_context() as body:
+            async with card_title.stream_context() as title:
+                async for chunk in stream:
+                    all_content += chunk
+                    deats = extract_image_details(all_content)
+                    await title.replace(deats.get("title"))
+                    await body.replace(image_details_ui(deats))
+                    yield chunk
 
     @reactive.effect
     @reactive.event(input.go)
     async def start_chat():
-        prompt = llm_prompt(input.style(), input.n_words())
-
-        # Start a new conversation
-        history = InMemoryChatMessageHistory()
-        nonlocal client
-        client = RunnableWithMessageHistory(prompt | model, lambda: history)
-        card_output = []
-
-        user_prompt = HumanMessage(
-            content=[
-                {
-                    "type": "text",
-                    "text": "Describe this image",
-                },
-                {"type": "image_url", "image_url": {"url": input.url()}},
-            ]
+        stream = await chat_client.stream_async(
+            "Describe this image",
+            content_image_url(input.url())
         )
-
-        stream = client.astream(user_prompt)
-        
-        stream2 = (update_card(chunk, card_output) async for chunk in stream)
-        await chat.append_message_stream(stream2)
+        await chat.append_message_stream(stream_wrapper(stream))
 
     # Allow the user to ask follow up questions
     @chat.on_user_submit
-    async def _():
-        print(chat.user_input())
-        user_message = HumanMessage(content=chat.user_input())
-        card_output = []
-        stream = client.astream(user_message)
+    async def _(user_input: str):
+        stream = await chat_client.stream_async(user_input)
+        await chat.append_message_stream(stream_wrapper(stream))
 
-        stream2 = (update_card(chunk, card_output) async for chunk in stream)
-        await chat.append_message_stream(stream2)
-            
-    @output
     @render.ui
-    @reactive.event(input.go)
     def chat_container():
-        return [
-            chat.ui()
-        ]
-
+        if input.go() == 0:
+            return "A chat will appear here once there is a description"
+        else:
+            return ui.chat_ui("chat")
+        
+    @reactive.effect
+    def _():
+        ui.update_sidebar(id="chat_sidebar", show=input.go() % 2 == 1)
```

Comment:
For my own edification what's the `% 2 == 1` at the end there?

Reply:
The idea was to show/hide the sidebar on odd/even clicks, but that's not a good idea, and this was never working. Instead of a nested sidebar, I updated to use offcanvas instead (see [30ed919](https://github.com/jonkeane/shinyImages/pull/1/commits/30ed91907c596b6b215e7e71cd28d532996a2bc4))

Reply:
👍